### ISSUE TITLE
NUnit.Console.exe sometimes returns with -100 (SocketException)

### ISSUE THREAD

**Issue by Lorilatschki** [nunit/nunit-console#1276]
We are using NUnit.Console.exe v3.16.0 and we are getting a -100 failure sporadically with a _System.Net.Sockets.SocketException_. 

NUnit v3.13.3
NUnit.Console.exe v3.16.0
NUnit.Extension.NUnitV2ResultWriter v3.6.1
NUnit.Extension.OverlayConfig v3.16.0.13 (custom extension built by us)

``Hint: I have adjusted all paths/names due to our company policy.``

We are running our tests like so (). The test-ids are collected by using the --explore console argument beforehand:
```cmd
c:\NUnit.ConsoleRunner.3.16.0\nunit3-console.exe "c:\tests\TestAssembly.dll" "/noheader" "/labels:before" "--trace=Debug" "--work=c:\RuntimeLogs" "--out=c:\RuntimeLogs\TestCaseA.txt" "/result=c:\XmlResults\TestCaseA.xml;format=nunit2" --where "id == 1-3702"
````

In case such an error occurs, only one trace file is being created. In that case InternalTrace.36104.log but no InternalTrace.36104.TestAssembly.dll.log. In case the error do not occur, the second log file is created.

> Unfortunately the mentioned nunit-agent_19228.log does **not** exist in the c:\RuntimeLogs folder. Any hint where it can be found?

InternalTrace.36104.log output:
```log
InternalTrace: Initializing at level Debug
10:11:26.307 Debug [ 1] ServiceManager: Added SettingsService
10:11:26.312 Debug [ 1] ServiceManager: Added RecentFilesService
10:11:26.312 Debug [ 1] ServiceManager: Added TestFilterService
10:11:26.314 Debug [ 1] ServiceManager: Added ExtensionService
10:11:26.314 Debug [ 1] ServiceManager: Added ProjectService
10:11:26.314 Debug [ 1] ServiceManager: Added RuntimeFrameworkService
10:11:26.417 Debug [ 1] ServiceManager: Added TestAgency
10:11:26.418 Debug [ 1] ServiceManager: Added DriverService
10:11:26.418 Debug [ 1] ServiceManager: Added ResultService
10:11:26.418 Debug [ 1] ServiceManager: Added DefaultTestRunnerFactory
10:11:26.419 Info  [ 1] ServiceManager: Initializing SettingsService
10:11:26.421 Info  [ 1] ServiceManager: Initializing RecentFilesService
10:11:26.423 Debug [ 1] ServiceManager: Request for service ISettings satisfied by SettingsService
10:11:26.424 Info  [ 1] ServiceManager: Initializing TestFilterService
10:11:26.424 Info  [ 1] ServiceManager: Initializing ExtensionService
10:11:26.428 Info  [ 1] ExtensionService: Scanning nunit.engine.core assembly for extension points
10:11:26.428 Info  [ 1] ExtensionService:   Found Path=/NUnit/Engine/NUnitV2Driver, Type=NUnit.Engine.Extensibility.IFrameworkDriver
10:11:26.453 Info  [ 1] ExtensionService: Scanning nunit.engine.api assembly for extension points
10:11:26.453 Info  [ 1] ExtensionService:   Found Path=/NUnit/Engine/TypeExtensions/IService, Type=NUnit.Engine.IService
10:11:26.453 Info  [ 1] ExtensionService:   Found Path=/NUnit/Engine/TypeExtensions/ITestEventListener, Type=NUnit.Engine.ITestEventListener
10:11:26.453 Info  [ 1] ExtensionService:   Found Path=/NUnit/Engine/TypeExtensions/IDriverFactory, Type=NUnit.Engine.Extensibility.IDriverFactory
10:11:26.453 Info  [ 1] ExtensionService:   Found Path=/NUnit/Engine/TypeExtensions/IProjectLoader, Type=NUnit.Engine.Extensibility.IProjectLoader
10:11:26.453 Info  [ 1] ExtensionService:   Found Path=/NUnit/Engine/TypeExtensions/IResultWriter, Type=NUnit.Engine.Extensibility.IResultWriter
10:11:26.461 Info  [ 1] ExtensionService: Processing file C:\NUnit.ConsoleRunner.3.16.0\nunit.console.nuget.addins
10:11:26.471 Info  [ 1] ExtensionService: Processing file C:\NUnit.ConsoleRunner.3.16.0\nunit.extensions.addins
10:11:26.511 Info  [ 1] ExtensionService: Scanning directory 'C:\NUnit.Extension.NUnitV2ResultWriter.3.6.1\tools' for extensions.
10:11:26.513 Info  [ 1] ExtensionService: Processing file C:\NUnit.Extension.NUnitV2ResultWriter.3.6.1\tools\net20.engine.addins
10:11:26.644 Info  [ 1] ExtensionService: Scanning directory 'C:\NUnit.Extension.OverlayConfig.3.16.0.132\net48\tools' for extensions.
10:11:26.644 Info  [ 1] ExtensionService: Processing file C:\NUnit.Extension.OverlayConfig.3.16.0.132\net48\tools\NUnit.Extension.OverlayConfig.addins
10:11:26.655 Info  [ 1] ExtensionService: Scanning C:\NUnit.Extension.NUnitV2ResultWriter.3.6.1\tools\net20\nunit-v2-result-writer.dll for Extensions
10:11:26.673 Info  [ 1] ExtensionService:   Found ExtensionAttribute on Type NUnit2XmlResultWriter
10:11:26.675 Info  [ 1] ExtensionService:         ExtensionProperty Format = nunit2
10:11:26.682 Info  [ 1] ExtensionService: Scanning C:\NUnit.Extension.OverlayConfig.3.16.0.132\net48\tools\NUnit.Extension.OverlayConfig.dll for Extensions
10:11:26.686 Info  [ 1] ExtensionService:   Found ExtensionAttribute on Type DriverFactory
10:11:26.686 Info  [ 1] ServiceManager: Initializing ProjectService
10:11:26.688 Debug [ 1] ServiceManager: Request for service ExtensionService satisfied by ExtensionService
10:11:26.690 Info  [ 1] ServiceManager: Initializing RuntimeFrameworkService
10:11:26.690 Info  [ 1] ServiceManager: Initializing TestAgency
10:11:26.691 Debug [ 1] ServiceManager: Request for service IRuntimeFrameworkService satisfied by RuntimeFrameworkService
10:11:26.728 Info  [ 1] ServiceManager: Initializing DriverService
10:11:26.730 Debug [ 1] ServiceManager: Request for service ExtensionService satisfied by ExtensionService
10:11:26.734 Info  [ 1] ServiceManager: Initializing ResultService
10:11:26.734 Debug [ 1] ServiceManager: Request for service ExtensionService satisfied by ExtensionService
10:11:26.735 Info  [ 1] ServiceManager: Initializing DefaultTestRunnerFactory
10:11:26.735 Debug [ 1] ServiceManager: Request for service IProjectService satisfied by ProjectService
10:11:26.735 Debug [ 1] ServiceManager: Request for service IResultService satisfied by ResultService
10:11:26.735 Debug [ 1] ServiceManager: Request for service ITestFilterService satisfied by TestFilterService
10:11:26.735 Debug [ 1] ServiceManager: Request for service IExtensionService satisfied by ExtensionService
10:11:26.822 Debug [ 1] ServiceManager: Request for service IProjectService satisfied by ProjectService
10:11:26.822 Debug [ 1] ServiceManager: Request for service ITestRunnerFactory satisfied by DefaultTestRunnerFactory
10:11:26.823 Debug [ 1] ServiceManager: Request for service IRuntimeFrameworkService satisfied by RuntimeFrameworkService
10:11:26.823 Debug [ 1] ServiceManager: Request for service ExtensionService satisfied by ExtensionService
10:11:27.289 Debug [ 1] RuntimeFrameworkService: Assembly C:\bin\TestAssembly.dll uses version 4.0.30319
10:11:27.289 Debug [ 1] RuntimeFrameworkService: Assembly C:\bin\TestAssembly.dll targets .NETFramework,Version=v4.8
10:11:27.294 Debug [ 1] RuntimeFrameworkService: Current framework is net-4.5
10:11:27.294 Debug [ 1] RuntimeFrameworkService: No specific framework requested for TestAssembly.dll
10:11:27.295 Debug [ 1] RuntimeFrameworkService: Test will use net-4.8 for TestAssembly.dll
10:11:27.295 Debug [ 1] RuntimeFrameworkService: Current framework is net-4.5
10:11:27.295 Debug [ 1] RuntimeFrameworkService: No specific framework requested for 
10:11:27.295 Debug [ 1] RuntimeFrameworkService: Test will use net-4.8 for 
10:11:27.297 Debug [ 1] ServiceManager: Request for service ITestRunnerFactory satisfied by DefaultTestRunnerFactory
10:11:27.297 Debug [ 1] ServiceManager: Request for service TestAgency satisfied by TestAgency
10:11:27.298 Info  [ 1] ProcessRunner: Running 
10:11:27.308 Debug [ 1] AgentProcess: TargetRuntime = net-4.8
10:11:27.311 Debug [ 1] AgentProcess: GetTestAgentExePath(net-4.8, False)
10:11:27.311 Debug [ 1] AgentProcess: Checking for agents at C:\NUnit.ConsoleRunner.3.16.0\agents
10:11:27.311 Debug [ 1] AgentProcess: Using nunit-agent at C:\NUnit.ConsoleRunner.3.16.0\agents\net462\nunit-agent.exe
10:11:27.485 Debug [ 1] TestAgency: Launched Agent process 19228 - see nunit-agent_19228.log
10:11:27.485 Debug [ 1] TestAgency: Command line: "C:\NUnit.ConsoleRunner.3.16.0\agents\net462\nunit-agent.exe" f7ac21be-2ad1-4a57-a7c8-1df7e39a7360 tcp://127.0.0.1:57615/TestAgency-93b26a3d-c6b2-4979-ac64-8419bbcfb5a6 --pid=36104 --trace=Debug --work=C:\RuntimeLogs
10:11:27.489 Debug [ 1] TestAgency: Waiting for agent {f7ac21be-2ad1-4a57-a7c8-1df7e39a7360} to register
10:11:28.100 Debug [ 1] TestAgency: Returning new agent {f7ac21be-2ad1-4a57-a7c8-1df7e39a7360}
10:13:05.513 Info  [ 1] ProcessRunner: Done running 
10:13:05.529 Info  [ 1] ProcessRunner: Unloading 
10:13:05.717 Debug [ 1] ProcessRunner: Stopping remote agent
10:13:05.967 Info  [ 1] ServiceManager: Stopping DefaultTestRunnerFactory
10:13:05.967 Info  [ 1] ServiceManager: Stopping ResultService
10:13:05.967 Info  [ 1] ServiceManager: Stopping DriverService
10:13:05.967 Info  [ 1] ServiceManager: Stopping TestAgency
10:13:05.967 Info  [ 1] ServiceManager: Stopping RuntimeFrameworkService
10:13:05.982 Info  [ 1] ServiceManager: Stopping ProjectService
10:13:05.982 Info  [ 1] ServiceManager: Stopping ExtensionService
10:13:05.982 Info  [ 1] ServiceManager: Stopping TestFilterService
10:13:05.982 Info  [ 1] ServiceManager: Stopping RecentFilesService
10:13:05.982 Info  [ 1] ServiceManager: Stopping SettingsService
```

Test output:
```log
Test console output:
Runtime Environment
   OS Version: Microsoft Windows NT 6.2.9200.0
   Runtime: .NET Framework CLR v4.0.30319.42000

Test Files
    C:\bin\TestAssembly.dll

Test Filters
    Where: id == 1-3702

System.Net.Sockets.SocketException : An existing connection was forcibly closed by the remote host

--SocketException
An existing connection was forcibly closed by the remote host

Server stack trace: 
   at System.Runtime.Remoting.Channels.SocketStream.Read(Byte[] buffer, Int32 offset, Int32 size)
   at System.Runtime.Remoting.Channels.SocketHandler.ReadFromSocket(Byte[] buffer, Int32 offset, Int32 count)
   at System.Runtime.Remoting.Channels.SocketHandler.Read(Byte[] buffer, Int32 offset, Int32 count)
   at System.Runtime.Remoting.Channels.Tcp.TcpFixedLengthReadingStream.Read(Byte[] buffer, Int32 offset, Int32 count)
   at System.IO.BinaryReader.ReadBytes(Int32 count)
   at System.Runtime.Serialization.Formatters.Binary.SerializationHeaderRecord.Read(__BinaryParser input)
   at System.Runtime.Serialization.Formatters.Binary.__BinaryParser.ReadSerializationHeaderRecord()
   at System.Runtime.Serialization.Formatters.Binary.__BinaryParser.Run()
   at System.Runtime.Serialization.Formatters.Binary.ObjectReader.Deserialize(HeaderHandler handler, __BinaryParser serParser, Boolean fCheck, Boolean isCrossAppDomain, IMethodCallMessage methodCallMessage)
   at System.Runtime.Serialization.Formatters.Binary.BinaryFormatter.Deserialize(Stream serializationStream, HeaderHandler handler, Boolean fCheck, Boolean isCrossAppDomain, IMethodCallMessage methodCallMessage)
   at System.Runtime.Remoting.Channels.CoreChannel.DeserializeBinaryResponseMessage(Stream inputStream, IMethodCallMessage reqMsg, Boolean bStrictBinding)
   at System.Runtime.Remoting.Channels.BinaryClientFormatterSink.SyncProcessMessage(IMessage msg)

Exception rethrown at [0]: 
   at System.Runtime.Remoting.Proxies.RealProxy.HandleReturnMessage(IMessage reqMsg, IMessage retMsg)
   at System.Runtime.Remoting.Proxies.RealProxy.PrivateInvoke(MessageData& msgData, Int32 type)
   at NUnit.Engine.ITestAgent.Stop()
   at NUnit.Engine.Runners.ProcessRunner.Dispose(Boolean disposing)
   at NUnit.Engine.Runners.AbstractTestRunner.Dispose()
   at NUnit.Engine.Runners.MasterTestRunner.Dispose(Boolean disposing)
   at NUnit.Engine.Runners.MasterTestRunner.Dispose()
   at NUnit.ConsoleRunner.ConsoleRunner.RunTests(TestPackage package, TestFilter filter)
   at NUnit.ConsoleRunner.Program.Main(String[] args)
```

**Comment by CharliePoole** [comment]
@Lorilatschki Did this process work before you upgraded to 3.16?

This may be related to #1275.

**Comment by Lorilatschki** [comment]
@CharliePoole this also fails sporadically with 3.15.0 and 3.15.2. Other versions we didn‘t tested yet because we recently changed from v2.x. 
I will have a look into your mentioned issue.

**Comment by CharliePoole** [comment]
The logs should all be found in the same directory. The agent could simply be failing before it creates the named log. Does the log appear when there is no failure?

**Comment by Lorilatschki** [comment]
I have checked the failing case scenario and indeed I found the second trace file by searching for PID of the nunit-agent.exe process. The trace output is:
```log
InternalTrace: Initializing at level Debug
10:11:27.720 Info  [ 1] ExtensionService: Scanning nunit.engine.core assembly for extension points
10:11:27.730 Info  [ 1] ExtensionService:   Found Path=/NUnit/Engine/NUnitV2Driver, Type=NUnit.Engine.Extensibility.IFrameworkDriver
10:11:27.774 Info  [ 1] ExtensionService: Scanning nunit.engine.api assembly for extension points
10:11:27.775 Info  [ 1] ExtensionService:   Found Path=/NUnit/Engine/TypeExtensions/IService, Type=NUnit.Engine.IService
10:11:27.775 Info  [ 1] ExtensionService:   Found Path=/NUnit/Engine/TypeExtensions/ITestEventListener, Type=NUnit.Engine.ITestEventListener
10:11:27.775 Info  [ 1] ExtensionService:   Found Path=/NUnit/Engine/TypeExtensions/IDriverFactory, Type=NUnit.Engine.Extensibility.IDriverFactory
10:11:27.775 Info  [ 1] ExtensionService:   Found Path=/NUnit/Engine/TypeExtensions/IProjectLoader, Type=NUnit.Engine.Extensibility.IProjectLoader
10:11:27.775 Info  [ 1] ExtensionService:   Found Path=/NUnit/Engine/TypeExtensions/IResultWriter, Type=NUnit.Engine.Extensibility.IResultWriter
10:11:27.783 Info  [ 1] ExtensionService: Processing file C:NUnitV3.x\bin\NUnit.ConsoleRunner.signed.3.16.0\agents\net462\nunit.agent.addins
10:11:27.788 Info  [ 1] ExtensionService: Scanning directory 'C:NUnitV3.x\bin\NUnit.ConsoleRunner.signed.3.16.0' for extensions.
10:11:27.789 Info  [ 1] ExtensionService: Processing file C:NUnitV3.x\bin\NUnit.ConsoleRunner.3.16.0\nunit.console.nuget.addins
10:11:27.792 Info  [ 1] ExtensionService: Processing file C:NUnitV3.x\bin\NUnit.ConsoleRunner.3.16.0\nunit.extensions.addins
10:11:27.800 Info  [ 1] ExtensionService: Scanning directory 'C:NUnitV3.x\bin\NUnit.Extension.NUnitV2ResultWriter.3.6.1\tools' for extensions.
10:11:27.801 Info  [ 1] ExtensionService: Processing file C:NUnitV3.x\bin\NUnit.Extension.NUnitV2ResultWriter.3.6.1\tools\net20.engine.addins
10:11:27.900 Info  [ 1] ExtensionService: Scanning directory 'C:NUnitV3.x\bin\NUnit.Extension.OverlayConfig.3.16.0.132\net48\tools' for extensions.
10:11:27.901 Info  [ 1] ExtensionService: Processing file C:NUnitV3.x\bin\NUnit.Extension.OverlayConfig.3.16.0.132\net48\tools\NUnit.Extension.OverlayConfig.addins
10:11:27.908 Info  [ 1] ExtensionService: Scanning C:NUnitV3.x\bin\NUnit.Extension.NUnitV2ResultWriter.3.6.1\tools\net20\nunit-v2-result-writer.dll for Extensions
10:11:27.924 Info  [ 1] ExtensionService:   Found ExtensionAttribute on Type NUnit2XmlResultWriter
10:11:27.925 Info  [ 1] ExtensionService:         ExtensionProperty Format = nunit2
10:11:27.932 Info  [ 1] ExtensionService: Scanning C:NUnitV3.x\bin\NUnit.Extension.OverlayConfig.3.16.0.132\net48\tools\NUnit.Extension.OverlayConfig.dll for Extensions
10:11:27.941 Info  [ 1] ExtensionService:   Found ExtensionAttribute on Type DriverFactory
10:11:27.956 Info  [ 1] TestAgentRemotingTransport: Agent starting
10:11:27.983 Info  [ 1] TestAgentRemotingTransport: Connecting to TestAgency at tcp://127.0.0.1:57615/TestAgency-93b26a3d-c6b2-4979-ac64-8419bbcfb5a6
10:11:28.084 Debug [ 1] TestAgentRemotingTransport: Registered agent with TestAgency
10:11:28.412 Info  [ 5] DomainManager: Creating application domain domain-
10:11:28.421 Debug [ 5] DriverService: Trying DriverFactory
10:11:28.598 Info  [ 5] NUnitFrameworkDriver: Loading TestAssembly.dll - see separate log file
10:11:29.376 Info  [ 5] NUnitFrameworkDriver: Loaded TestAssembly.dll
10:11:29.377 Info  [ 5] NUnitFrameworkDriver: Running TestAssembly.dll - see separate log file
10:13:05.732 Info  [ 5] TestAgentRemotingTransport: Stopping
10:13:05.732 Info  [47] TestAgentRemotingTransport: Waiting for messages to complete
10:13:05.732 Info  [47] TestAgentRemotingTransport: Attempting shut down channel
10:13:05.732 Info  [47] TestAgentRemotingTransport: Set stop signal
```
So no trace output other than Info/Debug 😞

**Comment by Lorilatschki** [comment]
@CharliePoole any idea how to proceed on that? We still getting this error regularly in a sporadic manner.

**Comment by CharliePoole** [comment]
@Lorilatschki Best best would be for you to create a non-intermittent repro, if possible.

There seems to be an associated issue here as well: the main log tells you to look for an `nunit_agent_*.log` file but it is actually created by the agent process with a different name.

FYI I'm actually no longer working on this project but I'm sure someone on the @nunit/engine-team will have to look at this.

**Comment by saiparth** [comment]
1. System.Net.Sockets.SocketException. I downloaded the zip file from 
I am also facing same issue.
Nunit 3.16 console
Nunit version is 3.13.3
```
<message><![CDATA[System.Net.Sockets.SocketException : An existing connection was forcibly closed by the remote host]]></message>
      <stack-trace><![CDATA[--SocketException
An existing connection was forcibly closed by the remote host
   at NUnit.Engine.Communication.Transports.Tcp.SocketReader.GetNextMessage() in C:\Users\charlie\dev\NUnit\nunit-console\src\NUnitEngine\nunit.engine.core\Communication\Transports\Tcp\SocketReader.cs:line 42
   at NUnit.Engine.Communication.Transports.Tcp.TestAgentTcpProxy.TestRunResult(ITestEventListener listener) in C:\Users\charlie\dev\NUnit\nunit-console\src\NUnitEngine\nunit.engine\Communication\Transports\Tcp\TestAgentTcpProxy.cs:line 116
   at NUnit.Engine.Runners.ProcessRunner.RunTests(ITestEventListener listener, TestFilter filter) in C:\Users\charlie\dev\NUnit\nunit-console\src\NUnitEngine\nunit.engine\Runners\ProcessRunner.cs:line 138
]]></stack-trace>
```
This issue is inconsistent and I am not sure where to find agent logs. 
2. In some cases below exception is displayed.
```
2023-01-25T03:00:21.0413325Z 1) Error : 
2023-01-25T03:00:21.0414701Z 
2023-01-25T03:00:21.0436746Z NUnit.Engine.NUnitEngineException : Unable to acquire remote process agent
2023-01-25T03:00:21.0438162Z 
2023-01-25T03:00:21.0460690Z --NUnitEngineException
2023-01-25T03:00:21.0462073Z 
2023-01-25T03:00:21.0484546Z Unable to acquire remote process agent
2023-01-25T03:00:21.0485932Z 
2023-01-25T03:00:21.0508134Z    at NUnit.Engine.Runners.ProcessRunner.CreateAgentAndRunnerIfNeeded() in C:\Users\charlie\dev\NUnit\nunit-console\src\NUnitEngine\nunit.engine\Runners\ProcessRunner.cs:line 253
2023-01-25T03:00:21.0509482Z 
2023-01-25T03:00:21.0531623Z    at NUnit.Engine.Runners.ProcessRunner.RunTests(ITestEventListener listener, TestFilter filter) in C:\Users\charlie\dev\NUnit\nunit-console\src\NUnitEngine\nunit.engine\Runners\ProcessRunner.cs:line 138
2023-01-25T03:00:21.0532968Z 
2023-01-25T03:00:21.0555224Z 
2023-01-25T03:00:21.0556657Z 
```

**Comment by Lorilatschki** [comment]
@saiparth do you get this error sporadically or constantly?

**Comment by saiparth** [comment]
@Lorilatschki 
sporadically.

**Comment by Lorilatschki** [comment]
@ nunit/engine-team do you have any idea to get more details about the issue?

**Comment by saiparth** [comment]
Hi @OsirisTerje 
Can you please check or inform any team member. We are facing lot of challenges due to this issue and not able to understand why this occurs. I believe below bugs are also somewhat similar.
https://github.com/nunit/nunit-console/issues/1189
https://github.com/nunit/nunit-console/issues/1190
https://github.com/nunit/nunit-console/issues/1262

**Comment by OsirisTerje** [comment]
@saiparth Will see if we can figure it out.

**Comment by saiparth** [comment]
Hi @OsirisTerje 
Any update on this issue?

**Comment by OsirisTerje** [comment]
@saiparth 

Sorry, no update yet.  These issues are hard to figure out.

1. Do you have any possibility to provide some kind of repro for this?  
2. I understand you're in the process of moving tests from NUnit V2 and up to V3.  Can you also use only the V3 report schema, and skip whatever is in V2?  
3. Any possibility for moving off the console and using dotnet test instead ?

**Comment by saiparth** [comment]
@OsirisTerje 
Thanks for replying. Sorry. I wont able to share any repo because why it happens I don't know. It works most of the time. Only thing I suspect is, during 1000+ test run, server will be slow, In that case NUnit agent is not able to register within timeout(30 seconds may be) and fails. Reducing worker count and Server restart once in a week is the workaround I am fallowing now. 

Framework was upgraded from 4.6 to 6. hence I wont be able to use any previous version. I cant because my reports have some dependency and I created utility for retry based on Nunit xml results.

**Comment by OsirisTerje** [comment]
@saiparth Sorry,  I didn't mean to share your production code.  But if you could create a small repro with some slow tests that would repro this.  You know the situation, so more straightforward for you to create something.  
It may very well be a timeout issue.  Have you looked at the source code and seen something that concerns you in that respect?

**Comment by saiparth** [comment]
I could not find anything so far, I will get back to you when I am able to reproduce the issue. there is no specific use-case, its 4000 tests running in app server everyday for over a week. Selenium and Rest sharp are used

**Comment by Lorilatschki** [comment]
Same on our site. We run about 2000 tests/fixtures each run (up to 128 in parallel dependend on our agent hardware set). Almost every second run we do have at least one failure saying

```
nunit-console.exe failed with error code -100
System.Net.Sockets.SocketException : An existing connection was forcibly closed by the remote host

--SocketException
An existing connection was forcibly closed by the remote host

Server stack trace: 
   at System.Runtime.Remoting.Channels.SocketStream.Read(Byte[] buffer, Int32 offset, Int32 size)
   at System.Runtime.Remoting.Channels.SocketHandler.ReadFromSocket(Byte[] buffer, Int32 offset, Int32 count)
   at System.Runtime.Remoting.Channels.SocketHandler.Read(Byte[] buffer, Int32 offset, Int32 count)
   at System.Runtime.Remoting.Channels.Tcp.TcpFixedLengthReadingStream.Read(Byte[] buffer, Int32 offset, Int32 count)
   at System.IO.BinaryReader.ReadBytes(Int32 count)
   at System.Runtime.Serialization.Formatters.Binary.SerializationHeaderRecord.Read(__BinaryParser input)
   at System.Runtime.Serialization.Formatters.Binary.__BinaryParser.ReadSerializationHeaderRecord()
   at System.Runtime.Serialization.Formatters.Binary.__BinaryParser.Run()
   at System.Runtime.Serialization.Formatters.Binary.ObjectReader.Deserialize(HeaderHandler handler, __BinaryParser serParser, Boolean fCheck, Boolean isCrossAppDomain, IMethodCallMessage methodCallMessage)
   at System.Runtime.Serialization.Formatters.Binary.BinaryFormatter.Deserialize(Stream serializationStream, HeaderHandler handler, Boolean fCheck, Boolean isCrossAppDomain, IMethodCallMessage methodCallMessage)
   at System.Runtime.Remoting.Channels.CoreChannel.DeserializeBinaryResponseMessage(Stream inputStream, IMethodCallMessage reqMsg, Boolean bStrictBinding)
   at System.Runtime.Remoting.Channels.BinaryClientFormatterSink.SyncProcessMessage(IMessage msg)

Exception rethrown at [0]: 
   at System.Runtime.Remoting.Proxies.RealProxy.HandleReturnMessage(IMessage reqMsg, IMessage retMsg)
   at System.Runtime.Remoting.Proxies.RealProxy.PrivateInvoke(MessageData& msgData, Int32 type)
   at NUnit.Engine.ITestAgent.Stop()
   at NUnit.Engine.Runners.ProcessRunner.Dispose(Boolean disposing)
   at NUnit.Engine.Runners.AbstractTestRunner.Dispose()
   at NUnit.Engine.Runners.MasterTestRunner.Dispose(Boolean disposing)
   at NUnit.Engine.Runners.MasterTestRunner.Dispose()
   at NUnit.ConsoleRunner.ConsoleRunner.RunTests(TestPackage package, TestFilter filter)
   at NUnit.ConsoleRunner.Program.Main(String[] args)
```
It is always a different test which fails.

**Comment by OsirisTerje** [comment]
@Lorilatschki 
1. Does this happen also when you run tests locally?
2. Can you provide a repro ?

**Comment by Lorilatschki** [comment]
@OsirisTerje 

> Does this happen also when you run tests locally?
Yes is also occurs locally!
> Can you provide a repro?
Due to our company policy we are not able to share this, sorry!

**Comment by OsirisTerje** [comment]
@Lorilatschki We don't want nor need your company code.  We need a repro from you, fake code, which is organized so that the error you report can be reproduced.  Then we have a better chance of fixing it.  
Since you confirm it also fails locally, then we have a much better chance of finding out what's going wrong here, but we need code that fails.

**Comment by Lorilatschki** [comment]
> Since you confirm it also fails locally, then we have a much better chance of finding out what's going wrong here, but we need code that fails

Since we do not know which is the root cause if the failure, it is hard to create a repo for that. I assume it is enough to run a banch of tests in parallel and from time to time this failure ocurrs. I can give it a try if I can easily create a repo to repoduce the issue.

> The other option is that you simply build the console code yourself, in debug mode, use that in your local build and then grab more information from that, possibly also being able to debug into the console/engine.

What do you mean grap more information from that? What information do you assume are part of the console code in debug mode rather thal release?

**Comment by Lorilatschki** [comment]
@OsirisTerje meanwhile we have enabled windows dump file creation for
```console
nunit3-console.exe
nunit-agent.exe
```
on all our agents. Hopefully we get a dump created in the -100 scenario. If so, would that help as well?

**Comment by JensLeible** [comment]
Hi there,

we were able to reproduce the issue on our end. This issue appears to be reproducible reliably when run with the -x86 flag for the nunit3-console. It's also flaky on our end when run without that.

I attached a [zipped test project](https://github.com/nunit/nunit-console/files/11342908/NUnitTestDummy.zip) to reproduce the issue. It contains a dummy test project targeted at net7.0 and a PowerShell file with the exact command we used along with the outcome that resulted in on our end, including the console output and an additional windows application error event with another interesting stack trace. Feel free to contact me for more information.

**Comment by JensLeible** [comment]
I'd like to emphasize that this is an important topic for us, since our CI build chain does not run reliably without it being fixed. The respective builds fail quite frequently. Would it be possible to provide a rough estimation on how long it will take to implement and release a fix. Thanks in advance!

**Comment by OsirisTerje** [comment]
I am sorry it is not possible to give any ETA for a fix.  The engine/console team is currently very small, and we're all occupied with the NUnit framework work.  The console/engine is also in a mixed state now, with version 3.15.X, 3.16.X and 4.X diverging.  We need to get those aligned.
I will try to get time to look into your repro asap, so that I at least get an idea of what is going on.  

That said, since you now have a repro, and the source code is fully available, you could make a try to dive into it, possibly also making some debugging changes along the way in the code, let it write out something useful, or even use the debugger to see if that helps.  That is more or less the same approach we would take here.

**Comment by OsirisTerje** [comment]
I just ran your repro, multiple times too.   I did use the 16.2 version which I already had installed, and it run with no errors.  Have you tried this on different machines?
Mine is a Windows 10 Pro, lots of RAM and CPU. 
![image](https://user-images.githubusercontent.com/203432/236493117-868c6309-fdcc-4085-9431-cbe8de1bd8d1.png)

**Comment by JensLeible** [comment]
Please have a look at the content of the _TestOutputOnMyMachine_ directory of the test project. The  _ProducedPowerShellConsole.log_ states that I was running **NUnit Console 3.16.3 (Release)**. OS Version is the same as in your log. Only difference is we are running Windows 10 Enterprise LTSC (Version 10.0.17763 Build 17763). But I would rather suspect the different console/enginge version being the reason for it not failing?

**Comment by OsirisTerje** [comment]
What happens using 3.16.2 on your side?

**Comment by JensLeible** [comment]
I just tested with 3.16.2 and it is also failing reliably on my machine.

An update on what I said about our build servers though: I tested the project there and it did not fail. However the Build chain did fail reliably. I cannot yet say what makes the difference there.

Anyway, your assumption is correct. It seems related to the Windows build.

**Comment by JensLeible** [comment]
Running against the test project on my machine with the current main branch (nunit4-console) of the nunit-console repo results in the following console output:

```log
NUnit Console 3.99.0.0-VSIDE (Debug)
Copyright (c) 2022 Charlie Poole, Rob Prouse
Monday, May 8, 2023 1:55:11 PM

Runtime Environment
    OS Version: Microsoft Windows NT 10.0.17763.0
    Runtime: .NET Framework CLR v4.0.30319.42000

Test Files
    .\NUnitTestDummy.dll

System.Net.Sockets.SocketException : An existing connection was forcibly closed by the remote host


Unhandled Exception: --SocketException
An existing connection was forcibly closed by the remote host
   at System.Net.Sockets.Socket.Send(Byte[] buffer)
   at NUnit.Engine.Communication.Transports.Tcp.TestAgentTcpProxy.SendCommandMessage(String command, Object[] arguments) in ~\nunit-console\src\NUnitEngine\nunit.engine\Communication\Transports\Tcp\TestAgentTcpProxy.cs:line 101
   at NUnit.Engine.Communication.Transports.Tcp.TestAgentTcpProxy.Stop() in ~\nunit-console\src\NUnitEngine\nunit.engine\Communication\Transports\Tcp\TestAgentTcpProxy.cs:line 43
   at NUnit.Engine.Communication.Transports.Remoting.TestAgentRemotingProxy.Stop() in ~\nunit-console\src\NUnitEngine\nunit.engine\Communication\Transports\Remoting\TestAgentRemotingProxy.cs:line 38
   at NUnit.Engine.Runners.ProcessRunner.Dispose(Boolean disposing) in ~\nunit-console\src\NUnitEngine\nunit.engine\Runners\ProcessRunner.cs:line 226
   at NUnit.Engine.Runners.TestEngineRunner.Dispose() in ~\nunit-console\src\NUnitEngine\nunit.engine\Runners\TestEngineRunner.cs:line 183
   at NUnit.Engine.Runners.MasterTestRunner.Dispose(Boolean disposing) in ~\nunit-console\src\NUnitEngine\nunit.engine\Runners\MasterTestRunner.cs:line 234
   at NUnit.Engine.Runners.MasterTestRunner.Dispose() in ~\nunit-console\src\NUnitEngine\nunit.engine\Runners\MasterTestRunner.cs:line 222
   at NUnit.ConsoleRunner.ConsoleRunner.RunTests(TestPackage package, TestFilter filter) in ~\nunit-console\src\NUnitConsole\nunit4-console\ConsoleRunner.cs:line 203
   at NUnit.ConsoleRunner.ConsoleRunner.Execute() in ~\nunit-console\src\NUnitConsole\nunit4-console\ConsoleRunner.cs:line 104
   at NUnit.ConsoleRunner.Program.Main(String[] args) in ~\nunit-console\src\NUnitConsole\nunit4-console\Program.cs:line 131

NUnit.Engine.NUnitEngineException: Remote test agent exited with non-zero exit code -532462766
   at NUnit.Engine.Services.TestAgency.OnAgentExit(Process process, Guid agentId) in ~\nunit-console\src\NUnitEngine\nunit.engine\Services\TestAgency.cs:line 159
   at NUnit.Engine.Services.TestAgency.<>c__DisplayClass14_0.<GetAgent>b__0(Object sender, EventArgs e) in ~\nunit-console\src\NUnitEngine\nunit.engine\Services\TestAgency.cs:line 74
   at System.Diagnostics.Process.OnExited()
   at System.Diagnostics.Process.RaiseOnExited()
   at System.Threading.ExecutionContext.RunInternal(ExecutionContext executionContext, ContextCallback callback, Object state, Boolean preserveSyncCtx)
   at System.Threading.ExecutionContext.Run(ExecutionContext executionContext, ContextCallback callback, Object state, Boolean preserveSyncCtx)
   at System.Threading._ThreadPoolWaitOrTimerCallback.PerformWaitOrTimerCallback(Object state, Boolean timedOut)
```

**Comment by OsirisTerje** [comment]
nunit4-console is not ready, but it obviously has an error.

**Comment by JensLeible** [comment]
What would be the correct branch to test with?

**Comment by JensLeible** [comment]
I just tested with said branch (version3).

Here's the console output with the nunit3-console:

```log
NUnit Console 3.99.0.0-VSIDE (Debug)
Copyright (c) 2022 Charlie Poole, Rob Prouse
Tuesday, May 9, 2023 11:30:55 AM

Runtime Environment
   OS Version: Microsoft Windows NT 6.2.9200.0
   Runtime: .NET Framework CLR v4.0.30319.42000

Test Files
    .\NUnitTestDummy.dll


Unhandled Exception: NUnit.Engine.NUnitEngineException: Remote test agent exited with non-zero exit code -532462766
   at NUnit.Engine.Services.TestAgency.OnAgentExit(Process process, Guid agentId) in D:\GitHub\nunit-console\src\NUnitEngine\nunit.engine\Services\TestAgency.cs:line 158
   at NUnit.Engine.Services.TestAgency.<>c__DisplayClass13_0.<GetAgent>b__0(Object sender, EventArgs e) in D:\GitHub\nunit-console\src\NUnitEngine\nunit.engine\Services\TestAgency.cs:line 73
   at System.Diagnostics.Process.OnExited()
   at System.Diagnostics.Process.RaiseOnExited()
   at System.Threading.ExecutionContext.RunInternal(ExecutionContext executionContext, ContextCallback callback, Object state, Boolean preserveSyncCtx)
   at System.Threading.ExecutionContext.Run(ExecutionContext executionContext, ContextCallback callback, Object state, Boolean preserveSyncCtx)
   at System.Threading._ThreadPoolWaitOrTimerCallback.PerformWaitOrTimerCallback(Object state, Boolean timedOut)

Errors, Failures and Warnings

1) Error : C:\Users\jleible\Desktop\NUnitTestDummy\bin\Debug\net7.0\NUnitTestDummy.dll
System.Net.Sockets.SocketException : An existing connection was forcibly closed by the remote host
--SocketException
An existing connection was forcibly closed by the remote host
   at System.Net.Sockets.Socket.Receive(Byte[] buffer)
   at NUnit.Engine.Communication.Transports.Tcp.SocketReader.GetNextMessage() in D:\GitHub\nunit-console\src\NUnitEngine\nunit.engine.core\Communication\Transports\Tcp\SocketReader.cs:line 42
   at NUnit.Engine.Communication.Transports.Tcp.TestAgentTcpProxy.TestRunResult(ITestEventListener listener) in D:\GitHub\nunit-console\src\NUnitEngine\nunit.engine\Communication\Transports\Tcp\TestAgentTcpProxy.cs:line 116
   at NUnit.Engine.Communication.Transports.Tcp.TestAgentTcpProxy.Run(ITestEventListener listener, TestFilter filter) in D:\GitHub\nunit-console\src\NUnitEngine\nunit.engine\Communication\Transports\Tcp\TestAgentTcpProxy.cs:line 73
   at NUnit.Engine.Runners.ProcessRunner.RunTests(ITestEventListener listener, TestFilter filter) in D:\GitHub\nunit-console\src\NUnitEngine\nunit.engine\Runners\ProcessRunner.cs:line 138

Test Run Summary
  Overall result: Failed
  Test Count: 0, Passed: 0, Failed: 0, Warnings: 0, Inconclusive: 0, Skipped: 0
  Start time: 2023-05-09 09:30:55Z
    End time: 2023-05-09 09:30:56Z
    Duration: 0.384 seconds

Results (nunit3) saved as .\TestOutput\results_N3.xml

NUnit.Engine.NUnitEngineUnloadException : Multiple exceptions encountered. Retrieve AggregatedExceptions property for more information
  ----> NUnit.Engine.NUnitEngineUnloadException : Agent Process was terminated successfully after error.
  ----> System.Net.Sockets.SocketException : An existing connection was forcibly closed by the remote host
```

With the _nunit3-netcore-console.exe_ however, the error does not occur:

```log
NUnit Console 3.99.0.0-VSIDE (Debug)
Copyright (c) 2022 Charlie Poole, Rob Prouse
Tuesday, May 9, 2023 11:38:28 AM

Runtime Environment
   OS Version: Microsoft Windows 10.0.17763
  Runtime: .NET 6.0.16

Test Files
    .\NUnitTestDummy.dll


Run Settings
    DisposeRunners: True
    SkipNonTestAssemblies: True
    WorkDirectory: C:\Users\jleible\Desktop\NUnitTestDummy\bin\Debug\net7.0
    NumberOfTestWorkers: 12

Test Run Summary
  Overall result: Passed
  Test Count: 1, Passed: 1, Failed: 0, Warnings: 0, Inconclusive: 0, Skipped: 0
  Start time: 2023-05-09 09:38:29Z
    End time: 2023-05-09 09:38:29Z
    Duration: 0.190 seconds

Results (nunit3) saved as .\TestOutput\results_N3.xml
```

**Comment by JensLeible** [comment]
Can I provide more information to help?

**Comment by OsirisTerje** [comment]
It is hard to see what we can do here with no repro that reproduces the problem.  You also found it worked on some machines, and not on yours and the build server.  I believe you are the closest person to find the cause here. 
The error code can indicate issues with security, with paths, nearly anything.  This might not be related to NUnit at all, but be more that NUnit is killed due to this.  
Since you can reproduce it locally, you can also alter the code around where it crashes, and dump more information out there, that is one possibility.  
Since it is the Unloading that fails, you could also catch the exception there, and just continue.  If that works, you could even raise a PR for that.

**Comment by JensLeible** [comment]
I have a quick side question: Are you aware of the fact that the NUnit.ConsoleRunner.NetCore nuget package on nuget.org holds a nunit3-netcore-console.dll instead of an executable? Since this could be a potential track to circumvent the current issue, I thought I'd ask that here. Follow up question would be: If you are aware, how should this package be used? I could not find any documentation about that.

![image](https://github.com/nunit/nunit-console/assets/11870267/003565af-8146-4957-9762-d60103c4c2d2)

**Comment by JensLeible** [comment]
Hi there,

one of our build servers managed to run the tests without problems. After some comparisons of the environment it turns out that the .NET SDK has been different. It failed on all build servers with .NET SDK 7.0.100. The successful one had .NET SDK 7.0.109 installed. After updating the other servers to .NET SDK 7.0.109, the tests run successfully there, too.

**Comment by Lorilatschki** [comment]
We kind of stick to net48 for the product we are facing the sporadic issues. 
For that we still seeing a lot of sporadic failures. So updating to the latest dotnet 7 sdk will not help!

**Comment by OsirisTerje** [comment]
@Lorilatschki Do you see the same sporadic issues if you use `dotnet test` instead of nunit.console ?

**Comment by Lorilatschki** [comment]
@OsirisTerje we try to setup several runs using the dotnet test instead of the nunit.console. We will update on this once we do have succeeded on that.

**Comment by Lorilatschki** [comment]
@OsirisTerje we are using the ``--where id=xxx`` filter to filter for concrete fixtures, this cannot be achived within the nunit3adapter since it only implement the category filter, does I got this right?

**Comment by OsirisTerje** [comment]
@Lorilatschki The adapter can also use the where filter, see https://docs.nunit.org/articles/vs-test-adapter/Tips-And-Tricks.html#where

**Comment by Lorilatschki** [comment]
@OsirisTerje unfortunately it doesn't work. The name of the test fixture is ``MyTestCase`` and the id is ``1-1012``. The id was determined via the ``--explore`` argument through the ``nunit3-console.exe``.

When I run 
```cmd
dotnet test "C:\temp\myTeset.dll" --test-adapter-path C:\temp\NUnit3TestAdapter\build\net48 -- NUnit.Where="id == 1-1012"
``` 
it states that there is no matching test ``No test is available``. When I use the name filter like
```cmd
dotnet test "C:\temp\myTeset.dll" --test-adapter-path C:\temp\NUnit3TestAdapter\build\net48 -- NUnit.Where="name == MyTestCase"
```
it works as expected and the test can be found and is executed. Any idea what I am doing wrong?

**Comment by OsirisTerje** [comment]
I would not trust the id to be the same. To determine the id in this case, use the dump flag and check if that actually is the same. https://docs.nunit.org/articles/vs-test-adapter/Tips-And-Tricks.html#dumpxmltestdiscovery-and-dumpxmltestresults  

But, do you actually need the id's ?

**Comment by Lorilatschki** [comment]
>To determine the id in this case, use the dump flag and check if that actually is the same.

I will give it a try.

>But, do you actually need the id's ?

This is hard to explain but let me try it shortly. Because we do have a lot of test dependency tools for our tests, they cannot be run in parallel out of the box, because some of them may not run together with some others. Since we do have a lot of tests and we wanted to take advantage of a parallel test execution we defined some attributes on our tests which are kind of blueprint how the test can run in parallel. 
Therefor we made some kind of test executer which collects all the tests (on fixture level) and than uses the nunit.console.exe to execute it. 
Since in nunit3 the explore was offered, we are determining the test fixtures using this command. Each and every testcase/fxiture has its own unique id, therefor we decided to use this in order to pass it into the where filter. 

We can change the filter under the hood to use the testname instead of the test id, but all that expense to run ``dotnet test`` instead of the console without knowing if this will fix this issue is kind of daring.

**Comment by Lorilatschki** [comment]
In fact the id's between the ``nunit.console3.exe --explore`` command and the NUnit3TestAdapter are not the same. 
Is there a way to explore/list the tests using the ``dotnet test`` command including the ids? I have tried the ``--list-tests`` argument but it only print out the test names.

**Comment by OsirisTerje** [comment]
If you're using `--list-tests` it should run the discovery, and then you would have the id's in the discovery dump file.

**Comment by Lorilatschki** [comment]
>UPDATE: Just tried it, and it didnt output more, so... you have to use the dump file

I can see no dump file in case I execute it with ``--list-tests``, where is it stored?

**Comment by OsirisTerje** [comment]
You should really try to get off that format :-)  
There actually is a NUnit2 result writer extension for the engine.  You might already be using that with the console.  I have never used it with the NUnit3 system, so I am on thin ice here.  But making that work with the adapter, I don't think that is possible without doing a code change.
The adapter has an extension point, which is used for calling into report extensions, but it is currently hardcoded to the nunit3 format.  It was the intention to allow this to be configurable, but since there are only two relevant formats, and one is deprecated (nunit2), it never got any traction. It is this code line https://github.com/nunit/nunit3-vs-adapter/blob/76b0d3dd48946f61e976c122db3f211fb1c7196a/src/NUnitTestAdapter/NUnitEngine/NUnitEngineAdapter.cs#L171
It might be that if the engine is set up to load the extension, and this line is changed to nunit2 it might work, but I would not bet too much on it.
The extension however, could be used to convert the NUnit3 to NUnit2 format, if we wrapped it in a console app.  See https://github.com/nunit/nunit-v2-result-writer  
There are also tools to convert between them like https://github.com/wooga/NUnit3-to-NUnit2-Format-Converter

The best is if you can check if it works to run the tests with dotnet test, and if that fails or not the same way.  I have never seen any errors like this with the dotnet test/adapter, so that is why I am asking.

**Comment by Lorilatschki** [comment]
Regarding the nunit2 format, we can get rid of it by adapting our overview generator to that.
>To set the result file, use the TestOutputXml command

Unfortunately this way you can only define the output folder rather than the output filename! Since we are running tests in parallel and the default name is the name of the test assembly (independend to what the fixture name is) there will be confilicts across the filenames. As I found in a separate [issue](https://github.com/nunit/docs/issues/592#issuecomment-1587591610), setting up the name of the result file is not possible yet within ``dotnet test``.

**Comment by OsirisTerje** [comment]
You can set the NewOutputXmlFileForEachRun to True and it will create new files if there already is a file with that name.  You will then get   `whatever.xml`, `whatever.1.xml`, `whatever.2.xml` and so on.

**Comment by Lorilatschki** [comment]
I don't think this is suiteable since the filenames should be predictable as we have some post processing of the files.
Currently we are storing all results in the same folder, if we use the ``NUnit.NewOutputXmlFileForEachRun=true`` setting, this will most likely fail since properly up to 80 tests are running in parallel and file sync's will only work safely in case they use a mutex, otherwise it can happen, that after test1 has not found a file named test.xml it assumes it is free but in the very same moment test2 creates such a file which will result in a file access violation.

So without the possibilty to define a own result file name, we cannot easily change to ``dotnet test``. 

We do have another issue which will arise when moving to ``dotnet test``. As I said, we do have some specific test tools within our test environment as well as a requirement, that for one of our module to test, it is necessary that the test host is signed with a specific key file. In our case, we do that by packing our own ``NUnit.ConsoleRunner`` by simply signing the ``nunit3-console.exe``. I know this sound wierd, but for this specific kind of tests we cannot change this requirement.

As a consequence we would love to get this fixed within the **NUnit.ConsoleRunner** itself, because switching to dotnet test will be huge effort and has some downsites due to the above mentioned aspects. 

_We much appreciate your support!_

**Comment by OsirisTerje** [comment]
To change things in the adapter is rather easy, so https://github.com/nunit/docs/issues/592 could be fixed.  Also including a mutex for the file creation (good point). 

Fixing the socket exception in the console though, is quite a different thing. We do need a working repro, which we don't have, and then we have some uncharted territory to go through there.  The NUnit.Console covers much more than the adapter based system, and we don't have anyone focused on the console right now. 

I'll try to collect together the different console issues that seems related and see if there are any good hints on what is actually causing this.  

Note also that you could build the console/engine code yourself try to get some more debugging information in there.  In particular in the NUnit.Engine.ITestAgent.Stop() method (this is an interface, so there are two possible implementation which could be the culprit) where the crash seems to originate.  Running a  debug  version might also give some more information.  

First, figure out which implementation is actually being used here, by adding some debug output in those two methods.  Then perhaps try a catch socketexception there.  
It would be much faster if you build this locally and create a nuget package locally that you can use for this purpose.

**Comment by Lorilatschki** [comment]
> ALTERNATIVELY: I could do some changes to get more debug info in, and release that out to our Myget feed. This would be more slow though, but the fact you have something where this is reproduced makes it interesting.

That sounds great, can you give a update once you have deployed such a version to the Myget feed?

> To change things in the adapter is rather easy, so https://github.com/nunit/docs/issues/592 could be fixed. Also including a mutex for the file creation (good point), alternatively added a guid to the filename to make them unique.

If I have some free time I can fix it and creating MRs out of it. Regarding the [naming issue](https://github.com/nunit/docs/issues/592), shall I add a new RunSettings property ``TestOutputXmlFileName`` where clients can setup a predefined name for the result file?

**Comment by Lorilatschki** [comment]
@OsirisTerje here are the two pull requests.

- [feat: enable custom test output result file](https://github.com/nunit/nunit3-vs-adapter/pull/1112)
- [refactor: use mutex for file determination](https://github.com/nunit/nunit3-vs-adapter/pull/1113)

**Comment by OsirisTerje** [comment]
@Lorilatschki Thanks a lot for this work!  It's really appreciated!  
I'll raise an issue in the adapter repo to connect these PRs, so that I can get it into the release notes.

**Comment by Lorilatschki** [comment]
>@Lorilatschki Thanks a lot for this work! It's really appreciated!

Same to me, your support is amazing!

**Comment by OsirisTerje** [comment]
The adapter with @Lorilatschki 's PRs are here:   https://www.myget.org/feed/nunit/package/nuget/NUnit3TestAdapter/4.5.0-dev-02467

**Comment by Lorilatschki** [comment]
@OsirisTerje since we figured out that using the ``dotnet test`` is not suiteable in our case, can I kindly ask you to provide a console package with further debug output according to? 
> ALTERNATIVELY: I could do some changes to get more debug info in, and release that out to our Myget feed. This would be more slow though, but the fact you have something where this is reproduced makes it interesting.

**Comment by OsirisTerje** [comment]
Yeah...  I am on travels right now, but back next week.  I'll try to get it out during that week.

**Comment by OsirisTerje** [comment]
@lsoft Yes, It might be the same issue. And yes, we're not sure how to handle this.  Also, BInaryFormatter is going away, so this has to be changed anyway. 
If you do have a viable repro solution, you could raise a new issue, referring to this, because this issue is now rather overloaded.

**Comment by LukeS5310** [comment]
Got this very similar issue with our test suite recently (occures after all tests had run and report is about to be generated):
```
Unhandled Exception: --SocketException
An existing connection was forcibly closed by the remote host
   at NUnit.Engine.Communication.Transports.Tcp.TestAgentTcpProxy.SendCommandMessage(String command, Object[] arguments)
   at NUnit.Engine.Runners.ProcessRunner.Dispose(Boolean disposing)
   at NUnit.Engine.Runners.AbstractTestRunner.Dispose()
   at NUnit.Engine.Runners.MasterTestRunner.Dispose(Boolean disposing)
   at NUnit.Engine.Runners.MasterTestRunner.Dispose()
   at NUnit.ConsoleRunner.ConsoleRunner.RunTests(TestPackage package, TestFilter filter)
   at NUnit.ConsoleRunner.Program.Main(String[] args)

NUnit.Engine.NUnitEngineException: Remote test agent exited with non-zero exit code -532462766
   at NUnit.Engine.Services.TestAgency.OnAgentExit(Process process, Guid agentId)
   at NUnit.Engine.Services.TestAgency.<>c__DisplayClass13_0.<GetAgent>b__0(Object sender, EventArgs e)
   at System.Diagnostics.Process.OnExited()
   at System.Diagnostics.Process.RaiseOnExited()
   at System.Threading.ExecutionContext.RunInternal(ExecutionContext executionContext, ContextCallback callback, Object state, Boolean preserveSyncCtx)
   at System.Threading.ExecutionContext.Run(ExecutionContext executionContext, ContextCallback callback, Object state, Boolean preserveSyncCtx)
   at System.Threading._ThreadPoolWaitOrTimerCallback.PerformWaitOrTimerCallback(Object state, Boolean timedOut)
```

stumbled upon this issue:  [#3446](https://github.com/nunit/nunit/issues/3446)
and followed advice of looking inside windows event viewer to find out that message exceeding maximum size (logs presumably):
```
Application: dotnet.exe
CoreCLR Version: 7.0.2024.26716
.NET Version: 7.0.20
Description: The process was terminated due to an unhandled exception.
Exception Info: System.Exception: Message is too big (148770846 bytes). Max allowed length is 134217728 bytes.
   at NUnit.Engine.Communication.Protocols.BinarySerializationProtocol.Encode(Object message) in d:\repos\NUnit\nunit-console\src\NUnitEngine\nunit.engine.core\Communication\Protocols\BinarySerializationProtocol.cs:line 40
   at NUnit.Engine.Communication.Transports.Tcp.TestAgentTcpTransport.SendResult(Object result) in d:\repos\NUnit\nunit-console\src\NUnitEngine\nunit.engine.core\Communication\Transports\Tcp\TestAgentTcpTransport.cs:line 126
   at NUnit.Engine.Communication.Transports.Tcp.TestAgentTcpTransport.CommandLoop() in d:\repos\NUnit\nunit-console\src\NUnitEngine\nunit.engine.core\Communication\Transports\Tcp\TestAgentTcpTransport.cs:line 106
```
Tweaking our logging output to be more compact did the trick. Hope this helps someone

**Comment by Leomana8** [comment]
Hi there, maybe our case could be helpful. 
We got the same exception "System.Net.Sockets.SocketException : An existing connection was forcibly closed by the remote host" when we upgraded project from Framework 4.6.2 to .Net 7.
This exception is misleading because it has no relation to a socket. This exception means that some kind of unhandled exception is thrown in your test suite. 
We ran our tests in debug mode with "nunit3-console.exe * --debug" (via Visual Studio connection) and we caught several different exceptions.
There were System.StackOverflowException and an exception from an unmanaged task (missing 'await').

**Comment by CharliePoole** [comment]
Note to self: confirm this on 3.19.2
